### PR TITLE
Symbolic encoding for multiples of pi.

### PR DESCRIPTION
**Pull request by ScottWe**
**Summary**

This pull request allows for the `QASMParser` to handle multiples of pi exactly.

**Details**

Exact circuit equivalence checking requires exact angles. The current implementation of `QASMParser` instead treats the symbolic constant `pi` as its numeric approximation `std::cos(-1)`. Such angles would be rejected by the equivalence checker in `src/python/verifier`. This commit resolves this issue by converting (most) supported OpenQASM expressions with `pi/n` to symbolic expressions with the `pi(n)` gate.

Of course, symbolic multiples of `pi` are not ideal for all use-cases. For this reason, the features described in this commit are controlled by a `symbolic_pi_` flag added to `QASMParser`. By default, this flag is disabled. It can be enabled at any time by calling the `void use_symbolic_pi(bool)` method. I did not add this to the constructor of `QASMParser` so as not to break pre-existing code.

**Main Changes**

- A `symbolic_pi` flag has been added to `QASMParser`. When this flag is set to `true`, all instances of `pi/n` are represented by the `pi(n)` gate, as opposed to a floating point literal.
- The `QASMParser` has been broken apart into a `QASMParser` and a `ParamParser`. The `ParamParser` handles all of the logic related to identifying symbolic pi expressions, and caching them appropriately (as concrete expressions are already cached).
- The `param_info` class can identify certain gates as "symbolic expressions". The list of these gates is defined by `is_symbolic_constant`. Currently, only the `pi` gate is treated as a symbolic constant. This stops the `get_new_param_expression_id` from optimizing away the pi expressions.

**Remark**

As mentioned above, the `symbolic_pi_` is disabled by default. Currently, it is only enabled in my test cases. This means that all existing executables in Quartz should behave the same as before.

### CODE REVIEW

**Review by xumingkuan — COMMENTED** [review]
Thanks!

File: src/quartz/parser/qasm_parser.h [hunk 1]
```diff
@@ -24,9 +24,26 @@ bool is_gate_string(const std::string &token, GateType &type);
...
+class ParamParser {
+ public:
+  ParamParser(Context *ctx, bool symbolic_pi)
+      : ctx_(ctx), symbolic_pi_(symbolic_pi) {}
+
+  int parse(std::string &token);
+
+ private:
+  int parse_deg(bool negative, ParamType p);
+  int parse_rad(bool negative, ParamType num, ParamType denom);
```
Comment (xumingkuan) [inline src/quartz/parser/qasm_parser.h:36]: I didn't understand the difference between `parse_deg` and `parse_rad`. If I understand correctly, this is not degree/radian, but `parse_number` and `parse_pi_expression`? It's probably also better to add a comment to show the results are `p` and `num * PI / denom` correspondingly.

Comment (ScottWe) [reply src/quartz/parser/qasm_parser.h]: I've tried clarifying this in one of the new commits.

**Review by xumingkuan — APPROVED** [review]
LGTM, thanks!

**Pull request merged**